### ISSUE TITLE
cmd.py command completion does not work with libedit

### ISSUE THREAD

**Issue by keeely** [python/cpython#102130]
# Bug report

As title, the Homebrew community has switched to libedit readline as of Python 3.11.  There exists a gnureadline package which allows one to continue using the gnu readline, but unfortunately cmd.py will continue to use the libedit one, and that doesn't always work properly.

# Your environment

Python 3.11 MacOs Ventura.

Here's some code that won't work with libedit readline:

```python
import cmd

class MyShell(cmd.Cmd):
    def do_command(self, _arg):
        print(f"running {self}")

MyShell().cmdloop()
```
Tab-completion appears to fail on Homebrew python 3.11, I suspect due to the libedit readline.  The code works fine with Homebrew python versions up to 3.10 which use gnu readline.  Also, if you install gnureadline and change all occurrences of `import readline` in cmd.py to `import gnureadline as readline` the cmd module works perfectly.

So in summary could we fix cmd.py to work correctly with libedit as well as gnu readline?  Or, failing that provide some fairly easy means of overriding the readline used by cmd.py without resorting to nasty monkey patching.

### Linked PRs
* gh-107738
* gh-107748

**Comment by ned-deily** [comment]
Thanks for your issue. Alas, the differences between GNU `readline` and the BSD-based `libedit` have been a source of confusion for many years for users of many commmand line tools, not just python-based ones. The [documentation for Python's `readline` module](https://docs.python.org/3/library/readline.html) discusses some of these differences in a note at the top of the page. In particular, it talks about the two different configuration files that are used, their different syntax, and what is needed to enable tab completion when using `libedit`.  For the record, one of the reasons why many distributions prefer to use `libedit` by default is its [more liberal licensing](https://en.wikipedia.org/wiki/GNU_Readline#Choice_of_the_GPL_as_GNU_Readline's_license): GNU `readline` is GPL-3 licensed.

Suggestions for improving the documentation or specific bugs when following the current suggestion in the docs would be welcomed. Otherwise, I'm not sure there's is any action to be taken here.

**Comment by keeely** [comment]
Thank you for that very helpful clue.  A quick google later, reveals I can change my code to this:

```
import cmd
import readline

if 'libedit' in readline.__doc__:
    print("Found libedit readline")
    readline.parse_and_bind("bind ^I rl_complete")
else:
    print("Found gnu readline")
    readline.parse_and_bind("tab: complete")


class MyShell(cmd.Cmd):
    def do_command(self, _arg):
        print(f"running {self}")

MyShell().cmdloop()
```
And after that the tab completion works for both libraries.  I have no idea why that works, but it does.  The libedit version isn't quite right, because the first press of tab does nothing but sound the bell, then the second (and subsequent) presses work as I'd expect.  A whole lot better than no completion, obviously.

It would be nice if Python abstracted this away to make cmd.py work out of the box for both libraries.

**Comment by carljm** [comment]
To me this looks like a legitimate bug report for the `cmd` module, which promises automatic  command completion in its docs, and has [code using the readline module to implement that](https://github.com/python/cpython/blob/main/Lib/cmd.py#L108), but that code appears to be GNU readline specific and doesn't work if `libedit` is used. (Unless we would want to say that the `cmd` module just doesn't support `libedit`, but that seems not ideal given its reasonably wide usage, and the fact that the author of a tool using `cmd` can't necessarily control whether their users have libedit or GNU readline installed.) The recommendations in the doc note for the `readline` module may provide pointers to how to fix `cmd`, but aren't in and of themselves an adequate solution for users of `cmd`: if you write a tool using `cmd`, it's not reasonable to tell all your users to put things in their `~/.editrc` to make your tool work either.

The ideal course of action here would be to figure out the right code needed in `cmd.py` to support libedit, and update `cmd.py` so that completion works out of the box with either underlying library.

If we achieve that, there may also be an argument for a follow-up feature where the `readline` module itself would incorporate some higher-level API that allows enabling completion without the client having to be aware of and check for libedit vs GNU readline (i.e. move the code to handle this from `cmd` to `readline`.) But I think that would be a separate issue from fixing the bug in `cmd`.

**Comment by carljm** [comment]
@hugovk I think maybe we should not tag this bug as `OS-mac` either; IMO the Homebrew switch to libedit is not really related to the actual bug here (libedit is used on other platforms too); it just means brew users will now be affected by the bug when they weren't before. What do you think?

**Comment by hugovk** [comment]
Makes sense, I've removed the label 👍

**Comment by keeely** [comment]
Whilst it would be great to fix cmd.py to make it perfect, I'd hate to chew through significant Python developer time to resolve all the issues with libedit readline.  There are at least two:  The first relating to history files, the second to tab completion.  Once people start using it in earnest we may find more.

Let's think for a moment who uses cmd.py.  It's for command-line interfaces, so for people who are comfortable with that, implying developers or technical people.  They won't think twice about installing a GNU dependency on their desktop machine.  Let's also think who has a problem with a GNU dependency.  That would be someone shipping a closed source product embedding Python with some kind of interactive configuration shell.  pfSense or something like it comes to mind, but these kinds of apps/devices usually ship with web interfaces.  It's all getting a bit niche.

I don't think it's full libedit support that's needed here, just a straightforward route for application maintainers (such as myself) to get their users working.  Adding the dependency on gnureadline is very easy for me.   I guess it breaks some fundamental rule in Python if we refer directly to a third-party package in cmd.py, so the next best thing is to pass it in.  If it's passed to the Cmd constructor, it could be used in preference to readline.

```
class Cmd:

    def __init__(self, completekey='tab', stdin=None, stdout=None, readline=None):
        self.readline = readline

# Everywhere readline is imported do something like
    def cmdloop(self, intro=None):
        ...
        if self.use_rawinput and self.completekey:
            try:
                if self.readline is None:
                    import readline
                else:
                    readline = self.readline
                self.old_completer = readline.get_completer()
                readline.set_completer(self.complete)
                readline.parse_and_bind(self.completekey+": complete")
            except ImportError:
                pass
```

**Comment by ned-deily** [comment]
Also, the differences you note here aren't limited to the python `cmd` module; they potentially affect any use of python's `readline` module and, for that matter, the Python interpreter's REPL shell. So if someone wants to go further than, say, making doc changes to the `cmd` module, I think it would be prudent to do some more investigation into how other distributions and users (python and non-python) of `libedit` vs ` GNU readline` have handled this (as @carljm points out, the issues here are not limited to one platform or distribution) and look at addressing them at the `readline` module level as much as possible.

**Comment by carljm** [comment]
> chew through significant Python developer time

Accepting that it's a valid bug is not a promise that core developers (or anyone else) will be able to prioritize fixing it, and it doesn't obligate anyone to use their time against their preferences. It just means that if someone comes along who is interested and motivated to fix it, they have some indication that a fix is desired and a good implementation of a fix is likely to be accepted.

> It's all getting a bit niche.

Lots of open issues on CPython are niche :) If they are too niche for anyone to care, they will eventually be closed due to inactivity / lack of interest. But lots of niche issues do find a motivated developer within that niche to fix them.

> There are at least two: The first relating to history files, the second to tab completion.

You've provided some details on the tab completion issue; can you say more about the history issue (i.e. what doesn't work, have you discovered any pointers about how to go about fixing it?) We could open a separate issue for it, but I'd be inclined for now to just broaden this issue to generally cover `cmd` compatibility with libedit, for both completion and history.

> the next best thing is to pass it in. If it's passed to the Cmd constructor, it could be used in preference to readline.

I don't think dependency-injecting an alternate `readline` module is a good long-term solution here. The standard library `readline` module should at least be good enough to support the standard library's own uses of `readline`. If someone really wants to inject a different readline module, it's already possible via `sys.modules`, as you've already found. (I don't think you necessarily have to bother with a metapath finder/loader for this; you can just import `gnureadline` early in your application's main module, before `cmd` is implemented, and then put it in `sys.modules['readline']`.)

**Comment by keeely** [comment]
That's a fair point.  It would be nice to pull in https://pypi.org/project/pyreadline3/ into the standard python distribution as well.  my install_requires for dealing with cmd.py requirements is now:
```
        "pyreadline3; os_name=='nt'",
        "gnureadline; sys_platform=='darwin' and python_version > '3.10.0'",
```

**Comment by dtrodrigues** [comment]
Some information about Python's history file format and libedit vs readline is in https://github.com/Homebrew/homebrew-core/pull/113811#issue-1419568255 

Another quirk referring to Python history search and libedit is https://github.com/python/cpython/issues/100610, which also probably should have the "OS-mac" label removed.

**Comment by keeely** [comment]
> > There are at least two: The first relating to history files, the second to tab completion.
> 
> You've provided some details on the tab completion issue; can you say more about the history issue (i.e. what doesn't work, have you discovered any pointers about how to go about fixing it?) We could open a separate issue for it, but I'd be inclined for now to just broaden this issue to generally cover `cmd` compatibility with libedit, for both completion and history.

I have a Mac and I've been through multiple OS updates, multiple Python updates, generally swapped between Homebrew and built-in Python so I don't understand the exact sequence of events, but attempting to read certain types of history files with libedit readline results in the error as described here:
https://stackoverflow.com/questions/29572484/os-x-can-anyone-explain-why-i-was-getting-an-os-error-in-my-pythonrc-py-file

I went back to Python with gnu readline, wrote some history, back to libedit, read the files and it wasn't that simple.  I didn't see the error.  Maybe newer versions of gnureadline stopped writing something incompatible to the files.

> (I don't think you necessarily have to bother with a metapath finder/loader for this; you can just import `gnureadline` early in your application's main module, before `cmd` is implemented, and then put it in `sys.modules['readline']`.)

Yes, I discovered this later on 👍

**Comment by Constantin1489** [comment]
After typing some codes like this in Python REPL,
```python
import sys
print('backslash \\011')
exit()
```

Libedit creates the history file below. GNU readline read it literally.
```text
_HiStOrY_V2_
import\040sys
print('backslash\040\134\134011')
exit()
```

[Libedit](https://github.com/NetBSD/src/tree/trunk/lib/libedit) ensures unicode encoding. Because ISO/IEC 10646 and Unicode have the same characters and numbers. See [Universal Coded Character Set - Differences from Unicode](https://en.wikipedia.org/wiki/Universal_Coded_Character_Set#Differences_from_Unicode)

`grep -i iso` in `lib/libedit` dir. See [NetBSD libedit repo](https://github.com/NetBSD/src/tree/trunk/lib/libedit). The entire repo is huge.

```text
chartype.h:43:#ifndef __STDC_ISO_10646__
chartype.h:45: * compatible, so ensure wchar_t indeed does ISO 10646 and not some other
chartype.h:47:	#error wchar_t must store ISO 10646 characters
chartype.h:52: * ref: ISO/IEC DTR 19769
```

In my case, `.python_history` of Libedit is written in UTF-8.

**To detect a history file from Libedit, check whether `_HiStOrY_V2_\n` is the first line.**

Related git commits of [Libedit](https://github.com/NetBSD/src/tree/trunk/lib/libedit) of NetBSD repo.
```text
_HiStOrY_V2_ : 84ee55f13e902adc52cfbbf77c12c35f3208d081 (Fri Feb 5 20:52:23 1999 +0000)
_HiStOrY_V1_ : 95b0e87b48a4e99f9ebc69c93c11e76ec6d1455b (Date:   Sun Oct 15 19:42:05 1995 +0000)
```

To improve the compatibility of history files between GNU readline and Libedit,
I suggest the pseudocode below.
```text
If there is '_HiStOrY_V2_\n' and GNU readline env,
ignore '_HiStOrY_V2_\n', read left file contents.
unicode-escape assume the encoding is latin-1.
therefore after decoding with unicode-escape, encode with latin-1 again, then decode it with utf-8.
and overwrite it with the python_history file.
after that, read the file with libedit.

Without '_HiStOrY_V2_\n' in first line, Libedit shred `.python_history`.
if there is no '_HiStOrY_V2_\n' in non empty `.python_history` and libedit env,
concatenate '_HiStOrY_V2_\n' with the file. and save it without modifying anything.
after that, read the file with libedit.
```

Because unicode-escape decode from latin-1 text according to Python [doc](https://docs.python.org/3/library/codecs.html#text-encodings).

# An example code.
Because readline Python module is written in c language, I can't write the c code.
```python
import os
import readline

def is_editline_history(first_line):
    if first_line == '_HiStOrY_V2_\n':
        return True
    return False

def is_readline():
    if 'libedit' in readline.__doc__:
        return False
    return True

def ensure_history_compatible(history_path='~/.python_history',
                            output='~/.python_history'):
    path = os.path.expanduser(history_path)
    output = os.path.expanduser(output)
    with open(path, 'rb') as f:
        history = first_line = f.readline()
        first_line_decoded = first_line.decode('unicode-escape')
        # If GNU readline reads Libedit history file.
        if is_editline_history(first_line_decoded) and is_readline():
            history = f.read()
            history = history.decode('unicode-escape').encode("latin-1")
        # If Libedit reads GNU readline history file.
        elif not is_editline_history(first_line_decoded) and not is_readline() :
            # Without '_HiStOrY_V2_\n', libedit shred the history file.
            # Luckily, without any octal character sequence converting, libedit read it as expected.
            history = '_HiStOrY_V2_\n'.encode() + history + f.read()

        # If no compatibility issue
        else:
            return

    with open(output, 'wb') as compatible_history:
        compatible_history.write(history)

if __name__ == "__main__":
    ensure_history_compatible()
```

REFERENCE
1. https://stackoverflow.com/a/29805750/20307768
2. https://stackoverflow.com/questions/52460662/how-to-decode-escaped-unicode-characters
3. https://stackoverflow.com/questions/4020539/process-escape-sequences-in-a-string-in-python
4. https://bugs.python.org/issue21331
5. https://stackoverflow.com/a/58829514
6. https://github.com/NetBSD/src/tree/trunk/lib/libedit
7. https://en.wikipedia.org/wiki/Unicode#Versions
8. https://en.wikipedia.org/wiki/Universal_Coded_Character_Set#Relationship_with_Unicode
9. https://en.wikipedia.org/wiki/Universal_Coded_Character_Set#Differences_from_Unicode


My env
```
# macOS
$ locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=

$ python                  
Python 3.11.4 (main, Jun 20 2023, 16:59:59) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import readline
>>> readline.__doc__
'Importing this module enables command line editing using libedit readline.'
```

```
#Rocky OS 9
$ locale
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=

$ python
Python 3.9.16 (main, May 29 2023, 00:00:00) 
[GCC 11.3.1 20221121 (Red Hat 11.3.1-4)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import readline
>>> readline.__doc__
'Importing this module enables command line editing using GNU readline.'
```

**Comment by Constantin1489** [comment]
Please don't think of me as arrogant. I don't want the backport for a different reason. Because editline support of python is not that good now. For example, `reverse-search-history` is not working for python REPL with editline(to enable `python:bind "^R" em-inc-search-prev` in `~/.editrc`). I don't want to give a false impression to advanced users... BTW, I started to assume that may be why the readline module is GNU readline interface...

**Comment by hugovk** [comment]
Triage: can this be closed or is there more to do?

**Comment by Constantin1489** [comment]
@hugovk https://github.com/python/cpython/issues/53279 is this relevant?

**Comment by encukou** [comment]
Yes, thanks for noticing.
Both can be closed.
